### PR TITLE
performance turning

### PR DESCRIPTION
**Pull request by ngaut**

### CODE REVIEW

File: kv/union_store.go [hunk 1]
```diff
@@ -46,15 +46,53 @@ type UnionStore struct {
...
+	if lw.mb == nil {
+		return nil, ErrNotExist
+	}
+
+	return lw.mb.Get(k)
+}
+
+func (lw *lazyMemBuffer) Set(key []byte, value []byte) error {
+	if lw.mb == nil {
+		lw.mb = p.Get().(MemBuffer)
```
Comment (shenli) [inline kv/union_store.go]: This pr LGTM.
But p is not a good name. Rename it to memdbPool is better. It is not a local variable but with single char name.

File: util/codec/bytes.go [hunk 2]
```diff
@@ -47,9 +47,12 @@ import (
 // EncodeBytes guarantees the encoded value is in ascending order for comparison,
 // The encoded value is >= SmallestNoneNilValue and < InfiniteValue.
 func EncodeBytes(b []byte, data []byte) []byte {
+	// Allocate more space to avoid unnecessary slice growing
+	bs := make([]byte, len(b), len(b)+len(data)+20)
```
Comment (siddontang) [inline util/codec/bytes.go]: sometimes we will alloc a proper b outer and reuse it to reduce memory allocation. like
```
b := make([]byte, 0, 1024)
b = EncodeBytes(b, data1)
b = EncodeBytes(b, data2)
```

I think this is more efficient than allocing inner each time.

File: util/codec/bytes.go [hunk 3]
```diff
@@ -78,7 +81,8 @@ func decodeBytes(b []byte, escapeFirst byte, escape byte, term byte) ([]byte, []
 		return nil, nil, errors.Errorf("insufficient bytes to decode value")
 	}
 
-	var r []byte
+	// Allocate more space to avoid unnecessary slice growing
+	r := make([]byte, 0, len(b)+20)
```
Comment (siddontang) [inline util/codec/bytes.go]: the b may be large (e.g, containing multi different encoded slices), so allocing with `len(b) + 20` may be not proper here.

**Comment by coocood** [comment]
Can you add benchmark for codec methods?
If the input `b` is large enough, the performance can be degraded

File: util/codec/bytes.go (hunk 2 above)
Comment (ngaut) [inline util/codec/bytes.go]: We write code rarely like that.  We just call EncodeBytes with nil as argument in most time.

File: util/codec/bytes.go (hunk 3 above)
Comment (ngaut) [inline util/codec/bytes.go]: Looks like allocate more 20 bytes doesn't hurt performance.

File: util/codec/bytes.go (hunk 2 above)
Comment (siddontang) [inline util/codec/bytes.go]: No, meta key codec use this way, and you can see the EncodeKey also uses it. 
So maybe the best is if b is nil, we alloc a proper capability.

File: util/codec/bytes.go (hunk 3 above)
Comment (siddontang) [inline util/codec/bytes.go]: what I mean is not extra 20 bytes, but the b here may contain multi large slices, so if we decode the first slice, we don't need to alloc whole len(b).

Comment (ngaut) [inline util/codec/bytes.go]: That's even more wired, do we use it like that(only decode the first slice)?

**Comment by ngaut** [comment]
PTAL

File: store/localstore/kv.go [hunk 4]
```diff
@@ -121,14 +111,17 @@ func (s *dbStore) UUID() string {
 }
 
 func (s *dbStore) GetSnapshot(ver kv.Version) (kv.MvccSnapshot, error) {
-	s.snapLock.RLock()
-	defer s.snapLock.RUnlock()
+	s.mu.RLock()
+	closed := s.closed
+	s.mu.RUnlock()
```
Comment (disksing) [inline store/localstore/kv.go]: What if concurrently call store.Close() just after release RLock?

Comment (ngaut) [inline store/localstore/kv.go]: It's ok. Reads and writes in the store will got an error after store closed.

Comment (disksing) [inline store/localstore/kv.go]: If that is true, we do not need the lock at all. Just use `if s.closed {...}`, behaviour is the same.

**Comment by ngaut** [comment]
PTAL @coocood @siddontang

File: store/localstore/compactor.go [hunk 5]
```diff
@@ -41,7 +41,7 @@ type localstoreCompactor struct {
 	recentKeys      map[string]struct{}
 	stopCh          chan struct{}
 	delCh           chan kv.EncodedKey
-	workerWaitGroup sync.WaitGroup
+	workerWaitGroup *sync.WaitGroup
```
Comment (siddontang) [inline store/localstore/compactor.go]: why use *sync.WaitGroup?

Comment (ngaut) [inline store/localstore/compactor.go]: To make go race happy.

File: store/localstore/snapshot.go [hunk 6]
```diff
@@ -31,15 +31,28 @@ var (
...
 type dbSnapshot struct {
-	store   *dbStore
-	db      engine.DB
-	rawIt   engine.Iterator
-	version kv.Version // transaction begin version
+	store    *dbStore
+	db       engine.DB
+	rawIt    engine.Iterator
+	version  kv.Version // transaction begin version
+	released bool
```
Comment (coocood) [inline store/localstore/snapshot.go]: Why add `released` field?

Comment (ngaut) [inline store/localstore/snapshot.go]: so call Release() twice has no side effects.

File: store/localstore/snapshot.go [hunk 7]
```diff
@@ -157,13 +170,17 @@ func (s *dbSnapshot) MvccRelease() {
 }
 
 func (s *dbSnapshot) Release() {
-	if s.rawIt == nil {
+	if s.released {
```
Comment (siddontang) [inline store/localstore/snapshot.go]: rawIt == nil can not check Released?

Comment (siddontang) [inline store/localstore/snapshot.go]: Got, ignore it.

**Comment by siddontang** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by zhiqiangxu** [comment]
[1, 2, 3, 4, 5, 6, 7, 8] -> [1, 2, 3, 4, 5, 6, 7, 8, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247]

BTW, why not encode it to [1, 2, 3, 4, 5, 6, 7, 8, 246], is it for historical/compatible reason?

**Pull request merged**